crates/bevy_app/src/sub_app.rs

```diff
@@ -0,0 +1,398 @@
+use crate::{App, First, InternedAppLabel, Plugin, Plugins, PluginsState, StateTransition};
+use bevy_ecs::{
+    prelude::*,
+    schedule::{
+        common_conditions::run_once as run_once_condition, run_enter_schedule,
+        InternedScheduleLabel, IntoSystemConfigs, IntoSystemSetConfigs, ScheduleLabel,
+    },
+};
+#[cfg(feature = "trace")]
+use bevy_utils::tracing::info_span;
+use bevy_utils::{default, HashMap, HashSet};
+use std::fmt::Debug;
+
+type ExtractFn = Box<dyn Fn(&mut World, &mut World) + Send>;
+
+#[derive(Default)]
+pub(crate) struct PluginStore {
+    pub(crate) registry: Vec<Box<dyn Plugin>>,
+    pub(crate) names: HashSet<String>,
+}
+
+/// A secondary application with its own [`World`]. These can run independently of each other.
+///
+/// These are useful for situations where certain processes (e.g. a render thread) need to be kept
+/// separate from the main application.
+pub struct SubApp {
+    /// The data of this application.
+    world: World,
+    /// Metadata for installed plugins.
+    pub(crate) plugins: PluginStore,
+    /// Panics if an update is attempted while plugins are building.
+    pub(crate) plugin_build_depth: usize,
+    pub(crate) plugins_state: PluginsState,
+    /// The schedule that will be run by [`update`](Self::update).
+    pub update_schedule: Option<InternedScheduleLabel>,
+    /// A function that gives mutable access to two app worlds. This is primarily
+    /// intended for copying data from the main world to secondary worlds.
+    extract: Option<ExtractFn>,
```

Comment:
This may be something for a future PR rather than this one, but I am curious whether it might be worth having a `Vec<SupAppLabel, ExtractFn>`?

Some potential situations I can think of for it include:
- having a `UI` world and the main world, and then having the render app extract from both
- having a simulation world that runs on a fixed update, which extracts  information from the main world for controls, but where the main world extracts from it for updating the displayed state

Reply:
That is definitely something for a different issue/PR.

---

Sub-apps are in a weird spot. I kinda don't expect them to last. Setting up the render world every frame is expensive enough that there's a lot of interest in figuring out how to do pipelined rendering without a second world.

If they stay, I think the future is storing other worlds as entities in the "main" world. In that case, users could write systems to copy data from one world to another or pass them between threads.

Reply:
That makes sense!

My brain is now jumping on some thoughts around namespacing entities as a potential way of handling things... or having "sub-app" marker components that are included by default in any filters...